### PR TITLE
Feat: Add ActiveProbe Support

### PR DESCRIPTION
**Pull request by cl1111**
Add activeprobe support

### CODE REVIEW

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 1]
```diff
@@ -36,19 +43,43 @@ import (
 	"github.com/everoute/everoute/pkg/constants"
 )
 
+const (
+	controllerName        = "activeprobe-controller"
+	externalIDIndex       = "externalIDIndex"
+	endpointExternalIDKey = "iface-id"
+	// Min and max data plane tag for traceflow. minTagNum is 7 (0b000111), maxTagNum is 59 (0b111011).
```
Comment (echkenluo) [inline pkg/controller/activeprobe/activeprobe_controller.go:46]: traceflow？

File: pkg/agent/controller/activeprobe/activeprobe_controller.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	}
+
+	apState := a.deleteActiveProbeByName(ap.Name)
+
+	if apState != nil {
+		err := a.DatapathManager.UninstallActiveProbeFlows(ovsbrName, apState.tag)
+		if err != nil {
+			klog.Errorf("Failed to uninstall Traceflow %s flows: %v", apState.name, err)
+		}
+	}
```
Comment (KeyongSun) [inline pkg/agent/controller/activeprobe/activeprobe_controller.go:270]: Traceflow? This logic can combine with deleteActiveProbeByname

File: pkg/agent/controller/activeprobe/packetin.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		reg5Val, err = getRegValue(matchers, activeProbeField)
+		if err != nil {
+			return state, tag, nil, err
+		}
+
+		switch {
+		case reg5Val == datapath.Tier1PolicyMatch && inPort == datapath.POLICY_TO_LOCAL_PORT:
+			telemetryTracePoint = activeprobev1alph1.IsolationPolicyEgress
+		case reg5Val == datapath.Tier2PolicyMatch && inPort == datapath.POLICY_TO_LOCAL_PORT:
+			telemetryTracePoint = activeprobev1alph1.FroensicPolicyEgress
```
Comment (KeyongSun) [inline pkg/agent/controller/activeprobe/packetin.go:128]: FroensicPolicyEgress -> ForensicPolicyEgress
and in activeprobe/v1alpha1 types.go FroensicPolicyEgresss -> ForenscPolicyEgress

File: pkg/agent/datapath/multiBridgeDatapath.go [hunk 4]
```diff
@@ -876,6 +928,85 @@ func (datapathManager *DpManager) RemoveEveroutePolicyRule(ruleID string, ruleNa
...
+
+func (datapathManager *DpManager) UninstallActiveProbeFlows(ovsbrName string, tag uint8) error {
+	for _, brName := range datapathManager.datapathConfig.ManagedVDSMap {
+		if brName == ovsbrName {
+			for _, flow := range datapathManager.ActiveProbeFlowMap[tag] {
+				if err := ofctrl.DeleteFlow(flow.Table, flow.Priority, flow.FlowID); err != nil {
+					return fmt.Errorf("failed to delete active probe flow: %v, error: %v", flow, err)
+				}
+			}
+			delete(datapathManager.ActiveProbeFlowMap, tag)
```
Comment (KeyongSun) [inline pkg/agent/datapath/multiBridgeDatapath.go:954]: if find the matched ovsbr at the end need to break the loop or return directly

File: pkg/agent/datapath/multiBridgeDatapath.go [hunk 5]
```diff
@@ -876,6 +928,85 @@ func (datapathManager *DpManager) RemoveEveroutePolicyRule(ruleID string, ruleNa
...
 
+func (datapathManager *DpManager) InstallActiveProbeFlows(ovsbrName string, tag uint8, ipDa *net.IP) error {
+	for vdsID, brName := range datapathManager.datapathConfig.ManagedVDSMap {
+		if brName == ovsbrName {
+			flowEntries, err := datapathManager.BridgeChainMap[vdsID][POLICY_BRIDGE_KEYWORD].(*PolicyBridge).InstallActiveProbeFlow(tag, ipDa)
+			if err != nil {
+				log.Errorf("Failed to install activeProbe flow. Err: %v", err)
+				return err
+			}
+			datapathManager.ActiveProbeFlowMap[tag] = flowEntries
```
Comment (KeyongSun) [inline pkg/agent/datapath/multiBridgeDatapath.go:939]: if matched the ovsbr need to break the loop or return.

File: pkg/agent/datapath/multiBridgeDatapath.go [hunk 6]
```diff
@@ -876,6 +928,85 @@ func (datapathManager *DpManager) RemoveEveroutePolicyRule(ruleID string, ruleNa
...
+			return sendActiveProbePacket(datapathManager.BridgeChainMap[vdsID][LOCAL_BRIDGE_KEYWORD].(*LocalBridge).OfSwitch,
+				tag, toOfctrlPacket(packet), inport, outport)
+		}
+	}
+	return nil
+}
+
+func toOfctrlPacket(packet Packet) *ofctrl.Packet {
+	return &ofctrl.Packet{
+		// TODO convert everoute packet to ofctrl packet
```
Comment (KeyongSun) [inline pkg/agent/datapath/multiBridgeDatapath.go:974]: What's the TODO meaning?

File: pkg/agent/datapath/policyBridge.go [hunk 7]
```diff
@@ -623,6 +663,56 @@ func (p *PolicyBridge) RemoveMicroSegmentRule(rule *EveroutePolicyRule) error {
 	return nil
 }
 
+func (p *PolicyBridge) InstallActiveProbeFlow(tag uint8, ipDa *net.IP) ([]*FlowEntry, error) {
```
Comment (KeyongSun) [inline pkg/agent/datapath/policyBridge.go:666]: ipDa -> ipData

File: pkg/agent/datapath/multiBridgeDatapath.go [hunk 8]
```diff
@@ -173,6 +179,49 @@ type Bridge interface {
...
+	if handler == nil {
+		log.Fatalf("Failed to register packetInHandler: register nil PacketInHandler is not allow")
+	}
+	if datapathManager.PacketInHandler != nil {
+		log.Fatalf("Failed to register PacketInHandler: datapathManager PacketInHandler already register")
+	}
+	datapathManager.PacketInHandler = handler
+}
+
+type Packet struct {
```
Comment (KeyongSun) [inline pkg/agent/datapath/multiBridgeDatapath.go:206]: Do we support the VLAN tag during the probe packet generate?

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+	objKey, ok := item.(k8stypes.NamespacedName)
+	if !ok {
+		r.syncQueue.Forget(item)
+		klog.Errorf("Activeprobe %v was not found in workqueue", objKey)
+		return
+	}
+
+	// TODO should support timeout and max retry
+	if err := r.syncActiveProbe(objKey); err == nil {
+		klog.Errorf("sync activeprobe  %v", objKey)
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:118]: Why use Errorf?

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+func (r *Reconciler) SyncActiveProbeWorker() {
+	item, shutdown := r.syncQueue.Get()
+	if shutdown {
+		return
+	}
+	defer r.syncQueue.Done(item)
+	// 1. lookup endpoint name from  endpoint agent
+
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:108]: format

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+	dstEndpointID := ctrltypes.ExternalID{
+		Name:  endpointExternalIDKey,
+		Value: dstEpExternalIDValue,
+	}
+
+	var epList = securityv1alpha1.EndpointList{}
+	err := r.Client.List(context.Background(), &epList)
+	if err != nil {
+		klog.Errorf("list endpoint: %s", err)
+		return nil
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:166]: err != nil why return nil?

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+func (r *Reconciler) deallocateTagForAP(ap *activeprobev1alph1.ActiveProbe) {
+	if ap.Status.Tag != 0 {
+		r.deallocateTag(ap.Name, ap.Status.Tag)
+	}
+}
+
+func (r *Reconciler) deallocateTag(name string, tag uint8) {
+	r.RunningActiveprobeMutex.Lock()
+	defer r.RunningActiveprobeMutex.Unlock()
+	if existingActiveProbeName, ok := r.RunningActiveprobe[tag]; ok {
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:214]: add a empty line before this

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+func (r *Reconciler) validateActiveProbe(ap *activeprobev1alph1.ActiveProbe) error {
+	return nil
+}
+
+func (r *Reconciler) updateActiveProbeStatus(ap *activeprobev1alph1.ActiveProbe,
+	state activeprobev1alph1.ActiveProbeState, reason string, tag uint8) error {
+	update := ap.DeepCopy()
+	update.Status.State = state
+	update.Status.Tag = tag
+	if reason != "" {
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:230]: add a empty line before this

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+	if existingActiveProbeName, ok := r.RunningActiveprobe[tag]; ok {
+		if name == existingActiveProbeName {
+			delete(r.RunningActiveprobe, tag)
+		}
+	}
+}
+
+func (r *Reconciler) validateActiveProbe(ap *activeprobev1alph1.ActiveProbe) error {
+	return nil
+}
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:223]: what's this logic here

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+		return err
+	}
+	if tag == 0 {
+		return nil
+	}
+
+	err = r.AddEndpointInfo(ap)
+	if err != nil {
+		return err
+	}
```
Comment (KeyongSun) [inline pkg/controller/activeprobe/activeprobe_controller.go:256]: Add some log before different error

File: pkg/controller/activeprobe/activeprobe_controller.go [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+		}
+	}
+	for i := minTagNum; i <= maxTagNum; i += tagStep {
+		if _, ok := r.RunningActiveprobe[i]; !ok {
+			r.RunningActiveprobe[i] = name
+			return i, nil
+		}
+	}
+	return 0, fmt.Errorf("number of on-going ActiveProve operations already reached the upper limit: %d", maxTagNum)
+}
```
Comment (wu-cl) [inline pkg/controller/activeprobe/activeprobe_controller.go:203]: RunningActiveprobe is stored in conntroller memory.
What will happen if controller HA to another one ?

**Comment by wu-cl** [comment]
`Supplement TCP message injection and parsing process`
two commits have same message, rename or squash them

File: pkg/agent/controller/activeprobe/activeprobe_controller.go (hunk 2 above)
Comment (cl1111) [reply pkg/agent/controller/activeprobe/activeprobe_controller.go]: has been modified

File: pkg/agent/controller/activeprobe/packetin.go (hunk 3 above)
Comment (cl1111) [reply pkg/agent/controller/activeprobe/packetin.go]: has been modified

File: pkg/agent/datapath/multiBridgeDatapath.go (hunk 6 above)
Comment (cl1111) [reply pkg/agent/datapath/multiBridgeDatapath.go]: has been modified

File: pkg/agent/datapath/multiBridgeDatapath.go (hunk 8 above)
Comment (cl1111) [reply pkg/agent/datapath/multiBridgeDatapath.go:206]: Temporarily not supported

File: pkg/controller/activeprobe/activeprobe_controller.go (hunk 9 above)
Comment (cl1111) [reply pkg/controller/activeprobe/activeprobe_controller.go]: has been modified

File: pkg/controller/activeprobe/activeprobe_controller.go (hunk 11 above)
Comment (cl1111) [reply pkg/controller/activeprobe/activeprobe_controller.go]: has been modified

File: pkg/controller/activeprobe/activeprobe_controller.go (hunk 12 above)
Comment (cl1111) [reply pkg/controller/activeprobe/activeprobe_controller.go:180]: has been modified

File: pkg/controller/activeprobe/activeprobe_controller.go (hunk 13 above)
Comment (cl1111) [reply pkg/controller/activeprobe/activeprobe_controller.go:198]: has been modified

File: pkg/controller/activeprobe/activeprobe_controller.go (hunk 15 above)
Comment (cl1111) [reply pkg/controller/activeprobe/activeprobe_controller.go]: has been modified

File: pkg/controller/activeprobe/activeprobe_controller.go (hunk 16 above)
Comment (cl1111) [reply pkg/controller/activeprobe/activeprobe_controller.go]: ok

**Review by KeyongSun — APPROVED** [review]

**Pull request merged**